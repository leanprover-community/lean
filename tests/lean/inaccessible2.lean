inductive imf {A B : Sort*} (f : A → B) : B → Sort*
| mk : ∀ (a : A), imf (f a)

definition inv_1 {A B : Sort*} (f : A → B) : ∀ (b : B), imf f b → A
| .(f .(a)) (imf.mk a)  := a  -- Error inaccessible annotation inside inaccessible annotation

definition inv_2 {A B : Sort*} (f : A → B) : ∀ (b : B), imf f b → A
| .(f a) (let x := (imf.mk a) in x)  := a  -- Error invalid occurrence of 'let' expression

definition inv_3 {A B : Sort*} (f : A → B) : ∀ (b : B), imf f b → A
| .(f a) ((λ (x : imf f b), x) (imf.mk a)) := a  -- Error invalid occurrence of 'lambda' expression

definition sym {A : Sort*} : ∀ a b : A, a = b → b = a
| .(a) .(a) (eq.refl a) := rfl -- Error `a` in eq.refl must be marked as inaccessible since it is an inductive datatype parameter

definition symm2 {A : Sort*} : ∀ a b : A, a = b → b = a
| _ _ rfl := rfl -- correct version
