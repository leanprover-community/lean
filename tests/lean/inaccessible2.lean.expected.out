inaccessible2.lean:5:8: error: invalid occurrence of 'inaccessible' annotation, it must only occur in patterns
inaccessible2.lean:8:10: error: invalid pattern, must be an application, constant, variable, type ascription, aliasing pattern or inaccessible term
inaccessible2.lean:8:40: error: unknown identifier 'a'
inaccessible2.lean:7:11: error: equation compiler failed (use 'set_option trace.eqn_compiler.elim_match true' for additional details)
inaccessible2.lean:11:11: error: invalid pattern, must be an application, constant, variable, type ascription, aliasing pattern or inaccessible term
inaccessible2.lean:11:39: error: don't know how to synthesize placeholder (a)
context:
A : Sort ?,
B : Sort ?,
f : A → B,
inv_3 : Π (b : B), imf f b → A
⊢ A
inaccessible2.lean:10:11: error: equation compiler failed (use 'set_option trace.eqn_compiler.elim_match true' for additional details)
inaccessible2.lean:14:21: error: don't know how to synthesize placeholder (a)
context:
A : Sort ?,
sym : ∀ (a b : A), a = b → b = a
⊢ A
inaccessible2.lean:13:11: error: equation compiler failed to create auxiliary declaration 'sym._main'
nested exception message:
type mismatch at application
  eq.rec (λ (a_1 : a = a) (H_2 : a_1 == eq.refl a), id_rhs (⁇ = ⁇) rfl)
term
  λ (a_1 : a = a) (H_2 : a_1 == eq.refl a), id_rhs (⁇ = ⁇) rfl
has type
  ∀ (a_1 : a = a), a_1 == _ → ⁇ = ⁇
but is expected to have type
  (λ (b : A), ∀ (a_1 : a = b), a_1 == _ → b = a) a
